### PR TITLE
Fixed edge cases for the bulk strings

### PR DESCRIPTION
**Pull request by barambani**

### CODE REVIEW

File: fs2/src/main/scala/laserdisc/fs2/RedisConnection.scala [hunk 1]
```diff
@@ -48,22 +48,34 @@ object RedisConnection {
 
       def framing: Pipe[F, Byte, CompleteFrame] = {
 
-        def go(stream: Stream[F, Byte], previous: Frame): Pull[F, CompleteFrame, Unit] =
+        def produceMultipleRight[A](xs: List[A]): Pull[F, A, Unit] =
```
Comment (barambani) [inline fs2/src/main/scala/laserdisc/fs2/RedisConnection.scala]: I'm not sure if there is a better way in fs2 for doing this. Any idea ?

**Comment by barambani** [comment]
With this the bulk string should be ok. I run hours of tests locally and the protocol looks implemented fully for bulk strings. 
Here are some numbers for different pooling configuration on a laptop with an imposed request concurrency
```
Server, Client 		 -> Same FixedThreadPool 16
AsynchronousChannelGroup -> FixedThreadPool 16
Scheduler Pool 		 -> FixedThreadPool 4

Request Concurrency = 16

Transactions:		      269780 hits
Availability:		      100.00 %
Elapsed time:		      600.90 secs
Data transferred:	      283.94 MB
Response time:		        0.02 secs
Transaction rate:	      448.96 trans/sec
Throughput:		        0.47 MB/sec
Concurrency:		       10.87
Successful transactions:      269780
Failed transactions:	           0
Longest transaction:	       20.34
Shortest transaction:	        0.00


Request Concurrency = 32

Transactions:		      279635 hits
Availability:		      100.00 %
Elapsed time:		      599.22 secs
Data transferred:	      294.31 MB
Response time:		        0.05 secs
Transaction rate:	      466.67 trans/sec
Throughput:		        0.49 MB/sec
Concurrency:		       21.31
Successful transactions:      279667
Failed transactions:	           3
Longest transaction:	        6.77
Shortest transaction:	        0.00
```
```
Server, Client 		 -> Same FixedThreadPool 32
AsynchronousChannelGroup -> FixedThreadPool 32
Scheduler Pool 		 -> FixedThreadPool 4

Request Concurrency = 16

Transactions:		      262371 hits
Availability:		      100.00 %
Elapsed time:		      600.02 secs
Data transferred:	      275.65 MB
Response time:		        0.02 secs
Transaction rate:	      437.27 trans/sec
Throughput:		        0.46 MB/sec
Concurrency:		        8.61
Successful transactions:      262386
Failed transactions:	           1
Longest transaction:	        1.45
Shortest transaction:	        0.00


Request Concurrency = 32

Transactions:		      269450 hits
Availability:		      100.00 %
Elapsed time:		      599.17 secs
Data transferred:	      283.26 MB
Response time:		        0.04 secs
Transaction rate:	      449.71 trans/sec
Throughput:		        0.47 MB/sec
Concurrency:		       20.21
Successful transactions:      269450
Failed transactions:	           1
Longest transaction:	       13.63
Shortest transaction:	        0.00
```
```
Server, Client 		 -> Same FixedThreadPool 16
AsynchronousChannelGroup -> FixedThreadPool 16
Scheduler Pool 		 -> FixedThreadPool 16

Request Concurrency = 16

Transactions:		      249982 hits
Availability:		      100.00 %
Elapsed time:		      599.23 secs
Data transferred:	      263.80 MB
Response time:		        0.02 secs
Transaction rate:	      417.17 trans/sec
Throughput:		        0.44 MB/sec
Concurrency:		        9.41
Successful transactions:      249982
Failed transactions:	           0
Longest transaction:	        5.46
Shortest transaction:	        0.00


Request Concurrency = 32

Transactions:		      282516 hits
Availability:		      100.00 %
Elapsed time:		      599.88 secs
Data transferred:	      296.76 MB
Response time:		        0.05 secs
Transaction rate:	      470.95 trans/sec
Throughput:		        0.49 MB/sec
Concurrency:		       22.97
Successful transactions:      282546
Failed transactions:	           2
Longest transaction:	        7.00
Shortest transaction:	        0.00
```
and with the siege `-b` option
```
Server 		         -> FixedThreadPool 16
Client 		         -> FixedThreadPool 16
AsynchronousChannelGroup -> FixedThreadPool 16
Scheduler Pool 		 -> FixedThreadPool 48

Transactions:		      288856 hits
Availability:		      100.00 %
Elapsed time:		      599.12 secs
Data transferred:	      621.20 MB
Response time:		        0.04 secs
Transaction rate:	      482.13 trans/sec
Throughput:		        1.04 MB/sec
Concurrency:		       17.04
Successful transactions:      288856
Failed transactions:	           1
Longest transaction:	       13.55
Shortest transaction:	        0.00
```
```
Server 		         -> FixedThreadPool 8
Client 		         -> FixedThreadPool 8
AsynchronousChannelGroup -> FixedThreadPool 8
Scheduler Pool 		 -> FixedThreadPool 4

Transactions:		      281418 hits
Availability:		      100.00 %
Elapsed time:		      599.21 secs
Data transferred:	      605.32 MB
Response time:		        0.04 secs
Transaction rate:	      469.65 trans/sec
Throughput:		        1.01 MB/sec
Concurrency:		       16.86
Successful transactions:      281442
Failed transactions:	           1
Longest transaction:	        2.26
Shortest transaction:	        0.00
```
```
Server 		         -> FixedThreadPool 8
Client 		         -> FixedThreadPool 8
AsynchronousChannelGroup -> FixedThreadPool 8
Scheduler Pool 		 -> FixedThreadPool 8

Transactions:		      278423 hits
Availability:		      100.00 %
Elapsed time:		      599.37 secs
Data transferred:	      598.85 MB
Response time:		        0.04 secs
Transaction rate:	      464.53 trans/sec
Throughput:		        1.00 MB/sec
Concurrency:		       19.64
Successful transactions:      278448
Failed transactions:	           1
Longest transaction:	       14.37
Shortest transaction:	        0.0
```
```
Server 		         -> Global
Client 		         -> FixedThreadPool 8
AsynchronousChannelGroup -> FixedThreadPool 8
Scheduler Pool 		 -> FixedThreadPool 8

Transactions:		      288685 hits
Availability:		      100.00 %
Elapsed time:		      599.97 secs
Data transferred:	      620.94 MB
Response time:		        0.03 secs
Transaction rate:	      481.17 trans/sec
Throughput:		        1.03 MB/sec
Concurrency:		       16.59
Successful transactions:      288685
Failed transactions:	           1
Longest transaction:	        1.99
Shortest transaction:	        0.00
```
```
Server 		         -> Global
Client 		         -> FixedThreadPool 16
AsynchronousChannelGroup -> FixedThreadPool 16
Scheduler Pool 		 -> FixedThreadPool 16

Transactions:		      277704 hits
Availability:		      100.00 %
Elapsed time:		      599.53 secs
Data transferred:	      597.26 MB
Response time:		        0.03 secs
Transaction rate:	      463.20 trans/sec
Throughput:		        1.00 MB/sec
Concurrency:		       14.87
Successful transactions:      277708
Failed transactions:	           2
Longest transaction:	        5.29
Shortest transaction:	        0.00
```

File: core/src/main/scala/laserdisc/protocol/RESP.scala [hunk 2]
```diff
@@ -254,7 +254,7 @@ sealed trait RESPCodecs { this: RESPBuilders =>
         override final def encode(bits: BitVector): Attempt[BitVector] = Attempt.successful(bits ++ crlf)
         override final def decode(bits: BitVector): Attempt[DecodeResult[BitVector]] =
           bits.bytes.indexOfSlice(crlfBytes, size) match {
-            case -1 => Attempt.failure(Err("Does not contain 'CRLF' termination bytes."))
+            case -1 => Attempt.failure(Err(s"Does not contain 'CRLF' termination bytes. Vector content: ${bits.toByteArray map (_.toChar) mkString ""}"))
```
Comment (sirocchj) [inline core/src/main/scala/laserdisc/protocol/RESP.scala]: maybe `new String(bits.toByteArray, UTF_8)`? Also, maybe worth replacing `\r\n` with escaped sequence like `.replaceAll("\r\n", "\\\\r\\\\n")`?

File: core/src/main/scala/laserdisc/protocol/RESP.scala [hunk 3]
```diff
@@ -393,48 +393,75 @@ sealed trait RESPFunctions { this: RESPCodecs =>
...
 
-      case (remainder, Multi) =>
+      case (remainder, RedisArray) =>
         arrayCodec.decode(remainder).fold(
           _ => Right(IncompleteVector),
           r => Right(CompleteAndDecoded(r))
         )
 
-      case (_, No)    => Right(CompleteVector)
+      case (_, SimpleString)
```
Comment (sirocchj) [inline core/src/main/scala/laserdisc/protocol/RESP.scala]: do we need to distinguish the 3?
also, what about `case (_, SimpleString | Integer | Error) => Right(CompleteVector)`?

File: fs2/src/main/scala/laserdisc/fs2/Frame.scala [hunk 4]
```diff
@@ -1,24 +1,48 @@
...
+
 import cats.syntax.either._
-import laserdisc.protocol.BitVectorDecoding._
+import laserdisc.protocol.BitVectorDecoding.{CompleteVector, _}
 import laserdisc.protocol.RESP
 import scodec.bits.BitVector
 
+import scala.annotation.tailrec
+
 sealed trait Frame extends Product with Serializable {
```
[the remark this answers is not in the document]
Comment (sirocchj) [reply fs2/src/main/scala/laserdisc/fs2/Frame.scala]: 👍

File: fs2/src/main/scala/laserdisc/fs2/Frame.scala [hunk 5]
```diff
@@ -1,24 +1,48 @@
 package laserdisc
 package fs2
 
-import _root_.fs2.Chunk
+import java.nio.ByteBuffer
+
 import cats.syntax.either._
-import laserdisc.protocol.BitVectorDecoding._
+import laserdisc.protocol.BitVectorDecoding.{CompleteVector, _}
```
Comment (sirocchj) [inline fs2/src/main/scala/laserdisc/fs2/Frame.scala]: is this change needed?

File: fs2/src/main/scala/laserdisc/fs2/Frame.scala [hunk 6]
```diff
@@ -1,24 +1,48 @@
...
-      case IncompleteVector      => Incomplete(bits, 0L)
-      case CompleteVector        => Complete(bits)
-    } leftMap (new Exception(_))
+    RESP.stateOf(bits) flatMap {
+      case MissingBits(n)              => Incomplete(bits, n).asRight
+      case CompleteAndDecoded(r)       => Decoded(r.value).asRight
+      case IncompleteVector            => Incomplete(bits, 0L).asRight
+      case CompleteVector              => Complete(bits).asRight
+      case CompleteWithRemainder(c, r) => consumeRemainder(MoreThanOne(Complete(c) :: Nil, r).asRight)
+    } leftMap (e => new Exception(s"Error: $e - Vector content: ${bits.toByteArray map (_.toChar) mkString ""}"))
```
Comment (sirocchj) [inline fs2/src/main/scala/laserdisc/fs2/Frame.scala]: same as first comment

File: fs2/src/main/scala/laserdisc/fs2/Frame.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (sirocchj) [reply fs2/src/main/scala/laserdisc/fs2/Frame.scala]: just careful if you do you'll lose the dep on cats, hence you won't have `EitherSyntax`

**Comment by barambani** [comment]
Just to add details, I attach here some charts about the GC and the HEAP during one of those tests. Also from the allocation point of view this should be a good start. 

screen shot 2018-08-07 at 11 46 47

File: fs2/src/main/scala/laserdisc/fs2/RedisConnection.scala (hunk 1 above)
Comment (sirocchj) [reply fs2/src/main/scala/laserdisc/fs2/RedisConnection.scala]: you want to emit the contents of `xs` (in reverse order) in here, right?
Not sure about perf but... what about `Pull.seq(xs.reverse)`? `Pull.seq` will actually convert it to `Vector` 😢

File: fs2/src/main/scala/laserdisc/fs2/Frame.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (barambani) [reply fs2/src/main/scala/laserdisc/fs2/Frame.scala]: I'm ok with that. here is used just for an embellishment

File: core/src/main/scala/laserdisc/protocol/RESP.scala (hunk 3 above)
Comment (barambani) [reply core/src/main/scala/laserdisc/protocol/RESP.scala]: I will refactor it. I will verify in another PR if the three of them are affected by the same issue I'm fixing here. If they are not needed I will remove them then.

File: core/src/main/scala/laserdisc/protocol/RESP.scala (hunk 2 above)
Comment (fiadliel) [reply core/src/main/scala/laserdisc/protocol/RESP.scala]: Bulk strings are binary-safe, so if you want a string that's always printable, maybe one of the representations (hex encoding?) from scodec?

Comment (fiadliel) [reply core/src/main/scala/laserdisc/protocol/RESP.scala]: Also, if the data is quite large, the most interesting information is the last few bytes, not the whole block.

File: fs2/src/main/scala/laserdisc/fs2/RedisConnection.scala (hunk 1 above)
Comment (sirocchj) [reply fs2/src/main/scala/laserdisc/fs2/RedisConnection.scala]: Sorry, I conflated two things: `Pull.seq` => `Pull.outputChunk(Chunk.seq(xs))` instead

File: fs2/src/main/scala/laserdisc/fs2/Frame.scala (hunk 5 above)
Comment (barambani) [reply fs2/src/main/scala/laserdisc/fs2/Frame.scala]: nope 👍
